### PR TITLE
fix: '_io.BytesIO' object has no attribute 'name' in paginator

### PR DESCRIPTION
**Pull request by yoggys**
## Summary

Fix issue with `discord/ext/pages/pagination.py` update function while using e.g. `io.BytesIO`.
Files that cannot be updated are skipped.

[Discord forum post](https://discord.com/channels/881207955029110855/1062602760899543070/1062602760899543070)

## Information

- [x] This PR fixes an issue.
- [ ] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed).
- [ ] This PR is **not** a code change (e.g. documentation, README, typehinting,
      examples, ...).

## Checklist

- [x] I have searched the open pull requests for duplicates.
- [x] If code changes were made then they have been tested.
  - [x] I have updated the documentation to reflect the changes.
- [ ] If `type: ignore` comments were used, a comment is also left explaining why.

### CODE REVIEW

**Comment by CompuGenius** [comment]
@yoggys, Thanks for this!

**Comment by Nzii3** [comment]
@yoggys Has the updated code been tested to make sure this works? If so, please provide screenshots and code usages so I can test this out myself with my test build.

**Comment by Nzii3** [comment]
Me and Bob are talking about the `update_files()` function, and it doesn’t seem to really update the files, since they are updated each time the page is changed.

**Comment by CompuGenius** [comment]
> @yoggys Has the updated code been tested to make sure this works? If so, please provide screenshots and code usages so I can test this out myself with my test build.

Here's a screenshot of the embed successfully having the proper image, which was sent with `io.BytesIO`. Without the above fix, the bot throws the error when trying to click to another page.

Screenshot_20230111_170431

Here's my code:
```py
with open(file_name, 'rb') as fp:
    data = io.BytesIO(fp.read())
file = discord.File(data, file_name)
```

**Comment by yoggys** [comment]
> @yoggys Has the updated code been tested to make sure this works? If so, please provide screenshots and code usages so I can test this out myself with my test build.

image

```py
import logging
import sys

import discord
from discord.ext import commands

logging.basicConfig(stream=sys.stdout, level=logging.INFO)

class Bot(commands.Bot):
    def __init__(self):
        super().__init__(command_prefix = '.', intents=discord.Intents.all())

bot = Bot()

@bot.slash_command()
async def teststringio(
    ctx: discord.ApplicationContext, 
): 
    from discord.ext import pages
    from io import StringIO
    test = discord.Embed(title="test", description="test")
    
    p = []
    for i in range(5):
        p.append(pages.Page(embeds=[test], files=[discord.File(StringIO(f"{i}"), f"test{i}.txt")]))
    paginator = pages.Paginator(pages=p, disable_on_timeout=True)
    await paginator.respond(ctx.interaction)
    
@bot.slash_command()
async def testreadingbuffer(
    ctx: discord.ApplicationContext, 
): 
    from discord.ext import pages
    test = discord.Embed(title="test", description="test")
    
    p = []
    for i in range(5):
        p.append(pages.Page(embeds=[test], files=[discord.File(f"test{i}.txt")]))
    paginator = pages.Paginator(pages=p, disable_on_timeout=True)
    await paginator.respond(ctx.interaction)

bot.run("token")
```

File: discord/ext/pages/pagination.py [hunk 1]
```diff
@@ -166,10 +167,12 @@ async def callback(self, interaction: Optional[discord.Interaction] = None):
         """
 
     def update_files(self) -> Optional[List[discord.File]]:
-        """Updates the files associated with the page by re-uploading them.
+        """If possible, it updates the files associated with the page by re-uploading them.
```
Comment (BobDotCom) [inline discord/ext/pages/pagination.py:170]:
```suggestion
        """Re-opens and reads new file contents for local files if they were updated.
```

**Review by BobDotCom — CHANGES_REQUESTED** [review]
Please check the suggestion I added. Also, please add a changelog entry.

File: discord/ext/pages/pagination.py [hunk 2]
```diff
@@ -166,10 +167,12 @@ async def callback(self, interaction: Optional[discord.Interaction] = None):
         """
 
     def update_files(self) -> Optional[List[discord.File]]:
```
Comment (BobDotCom) [inline discord/ext/pages/pagination.py:169]: While we're at it, let's update the typehint to the new syntax.
```suggestion
    def update_files(self) -> list[discord.File] | None:
```

**Comment by yoggys** [comment]
whops, wrong section - i moved it to `Changed`

File: CHANGELOG.md [hunk 3]
```diff
@@ -28,6 +28,9 @@ These changes are available on the `master` branch, but have not yet been releas
 
 - Changed `EmbeddedActivity` values to update accordingly with the new activities.
   ([#1859](https://github.com/Pycord-Development/pycord/pull/1859))
+- Changed the returned value, docstring and fixed a bug where types other than
+  `io.BufferedReader` in `Page.update_files` were throwing `AttributeError`.
+  ([#1869](https://github.com/Pycord-Development/pycord/pull/1869))
```
Comment (BobDotCom) [inline CHANGELOG.md:33]: We don't need to list documentation changes, and the typehint updates are stylistic - the new value is equivalent to the old one - so we don't need to list it either. That just leaves the fix, so we can put it under the "fixed" section (see https://keepachangelog.com/en/1.0.0/#types).
```suggestion

### Fixed
- Fixed a bug in `Page.update_files` where file objects stored in memory were causing an `AttributeError`. ([#1869](https://github.com/Pycord-Development/pycord/pull/1869))
```

**Review by BobDotCom — APPROVED** [review]

File: CHANGELOG.md (hunk 3 above)
Comment (yoggys) [reply CHANGELOG.md]: fine, thanks!

File: discord/ext/pages/pagination.py [hunk 4]
```diff
@@ -21,7 +21,10 @@
 FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 DEALINGS IN THE SOFTWARE.
 """
-from typing import Dict, List, Optional, Union
```
Comment (Om1609) [inline discord/ext/pages/pagination.py:24]: Py-cord supports Python 3.8+ but typehinting without the `typing` module was introduced only in 3.9

Comment (yoggys) [reply discord/ext/pages/pagination.py:24]: > Py-cord supports Python 3.8+ but typehinting without the `typing` module was introduced only in 3.9

hm @pre-commit-ci commited this after your commit @BobDotCom i dont know if this change should be rolled back in this case or just bot commit

Comment (BobDotCom) [reply discord/ext/pages/pagination.py:24]: The `from __future__ import annotations` import fixes this. See [the python documentation](https://docs.python.org/3/library/__future__.html) and [my message in the discord server](https://canary.discord.com/channels/881207955029110855/1062602760899543070/1062800078277972058) for more info.

**Review by Lulalaby — APPROVED** [review]
Well then, lgtm

here we go~

**Pull request merged**